File: src/test/java/com/exasol/cloudetl/kafka/KafkaTopicMetadataReaderTest.java
package com.exasol.cloudetl.kafka;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.containsString;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.Arrays;

import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.PartitionInfo;
import org.apache.kafka.common.errors.*;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class KafkaTopicMetadataReaderTest {
    private static final String TOPIC = "orders";
    @Mock
    KafkaConsumer<Void, Void> consumerMock;

    @Test
    void readsTopicPartitionsAndClosesConsumer() {
        when(consumerMock.partitionsFor(TOPIC)).thenReturn(Arrays.asList(new PartitionInfo(TOPIC, 0, null, null, null),
                new PartitionInfo(TOPIC, 2, null, null, null)));

        final var partitions = KafkaTopicMetadataReader.getTopicPartitions(consumerMock, TOPIC);

        assertAll(() -> assertEquals(Arrays.asList(0, 2), partitions),
                () -> verify(consumerMock).close());
    }

    @Test
    void mapsTimeoutErrorsToConnectorException() {
        assertMappedException(new TimeoutException("timed out"), "E-KCE-24",
                "Timeout trying to connect to Kafka brokers");
    }

    @Test
    void mapsAuthorizationErrorsToConnectorException() {
        assertMappedException(new AuthorizationException("missing ACL"), "E-KCE-25",
                "does not have access to read the given topic");
    }

    @Test
    void mapsAuthenticationErrorsToConnectorException() {
        assertMappedException(new AuthenticationException("bad credentials"), "E-KCE-26",
                "Failed to authenticate to the Kafka cluster");
    }

    @Test
    void mapsUnexpectedErrorsToConnectorException() {
        when(consumerMock.partitionsFor(TOPIC)).thenThrow(new IllegalStateException("boom"));

        final KafkaConnectorException thrown = assertThrows(KafkaConnectorException.class,
                () -> KafkaTopicMetadataReader.getTopicPartitions(consumerMock, TOPIC));

        assertAll(() -> assertThat(thrown.getMessage(), containsString("F-KCE-27")),
                () -> assertInstanceOf(IllegalStateException.class, thrown.getCause()),
                () -> verify(consumerMock).close());
    }

    private void assertMappedException(final RuntimeException exception, final String errorCode,
            final String messagePart) {
        when(consumerMock.partitionsFor(TOPIC)).thenThrow(exception);

        final KafkaConnectorException thrown = assertThrows(KafkaConnectorException.class,
                () -> KafkaTopicMetadataReader.getTopicPartitions(consumerMock, TOPIC));

        assertAll(() -> assertThat(thrown.getMessage(), containsString(errorCode)),
                () -> assertThat(thrown.getMessage(), containsString(messagePart)),
                () -> verify(consumerMock).close());
    }
}

File: src/test/scala/com/exasol/cloudetl/kafka/KafkaTopicMetadataReaderTest.scala
package com.exasol.cloudetl.kafka

import java.util.Arrays

import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.common.PartitionInfo
import org.apache.kafka.common.errors.AuthenticationException
import org.apache.kafka.common.errors.AuthorizationException
import org.apache.kafka.common.errors.TimeoutException
import org.mockito.Mockito.verify
import org.mockito.Mockito.when
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.mockito.MockitoSugar

class KafkaTopicMetadataReaderTest extends AnyFunSuite with MockitoSugar {

  private[this] val topic = "orders"

  test("must read topic partitions and close the consumer") {
    val consumer = mock[KafkaConsumer[Void, Void]]
    when(consumer.partitionsFor(topic)).thenReturn(
      Arrays.asList(
        new PartitionInfo(topic, 0, null, null, null),
        new PartitionInfo(topic, 2, null, null, null)
      )
    )

    val partitions = KafkaTopicMetadataReader.getTopicPartitions(consumer, topic)

    assert(partitions === List(0, 2))
    verify(consumer).close()
  }

  test("must map timeout errors to a connector exception") {
    val consumer = mock[KafkaConsumer[Void, Void]]
    when(consumer.partitionsFor(topic)).thenThrow(new TimeoutException("timed out"))

    val thrown = intercept[KafkaConnectorException] {
      KafkaTopicMetadataReader.getTopicPartitions(consumer, topic)
    }

    assert(thrown.getMessage.contains("E-KCE-24"))
    assert(thrown.getMessage.contains("Timeout trying to connect to Kafka brokers"))
    verify(consumer).close()
  }

  test("must map authorization errors to a connector exception") {
    val consumer = mock[KafkaConsumer[Void, Void]]
    when(consumer.partitionsFor(topic)).thenThrow(new AuthorizationException("missing ACL"))

    val thrown = intercept[KafkaConnectorException] {
      KafkaTopicMetadataReader.getTopicPartitions(consumer, topic)
    }

    assert(thrown.getMessage.contains("E-KCE-25"))
    assert(thrown.getMessage.contains("does not have access to read the given topic"))
    verify(consumer).close()
  }

  test("must map authentication errors to a connector exception") {
    val consumer = mock[KafkaConsumer[Void, Void]]
    when(consumer.partitionsFor(topic)).thenThrow(new AuthenticationException("bad credentials"))

    val thrown = intercept[KafkaConnectorException] {
      KafkaTopicMetadataReader.getTopicPartitions(consumer, topic)
    }

    assert(thrown.getMessage.contains("E-KCE-26"))
    assert(thrown.getMessage.contains("Failed to authenticate to the Kafka cluster"))
    verify(consumer).close()
  }

  test("must map unexpected errors to a connector exception") {
    val consumer = mock[KafkaConsumer[Void, Void]]
    when(consumer.partitionsFor(topic)).thenThrow(new IllegalStateException("boom"))

    val thrown = intercept[KafkaConnectorException] {
      KafkaTopicMetadataReader.getTopicPartitions(consumer, topic)
    }

    assert(thrown.getMessage.contains("F-KCE-27"))
    assert(thrown.getCause.isInstanceOf[IllegalStateException])
    verify(consumer).close()
  }
}
